src/Pool.sol

```diff
@@ -345,6 +363,50 @@ abstract contract Pool {
         receiver.nextCollectedCycle = cycle;
     }
 
+    /// @notice Gives funds to the receiver.
+    /// The receiver can collect them immediately.
+    /// @param giverAddr The address of the giver
+    /// @param receiverAddr The receiver
+    /// @param amt The given amount
+    function _giveInternal(
+        address giverAddr,
+        address receiverAddr,
+        uint128 amt
+    ) internal {
+        emit Given(giverAddr, receiverAddr, amt);
+        _giveFromAnyGiver(giverAddr, receiverAddr, amt);
```

Comment:
I would switch the order. In case a of a failed transaction (missing approval) it will be a bit cheaper if the emit is after the potential revert.

Reply:
I wouldn't care too much about revert gas cost, it's a rare case when a transaction is submitted despite being reverting.

What I'm trying to achieve here is putting `_transfer` as the very last thing the contract does, because it's a call to an external contract. In this particular function it probably doesn't matter much, but generally it's a good habit, which can prevent nasty scenarios like a re-entry attack.

Reply:
> I wouldn't care too much about revert gas cost, it's a rare case when a transaction is submitted despite being reverting.
It is not so rare, especially if gas prices are high.  You might have pending transaction and you resubmit a new one. Metamask won't revert because approval is here both not for both, etc.

But not a lot of savings, I agree. 

I agree with the re-entrance attack. Normally the rule is bookkeeping first afterwards the payout. 


In our case we transfer the DAI to the contract so maybe in that case the bookkeeping after we received the DAI.